packages/viewer/src/scenes/operations.ts

```diff
@@ -0,0 +1,69 @@
+import { ColorMaterial } from './colorMaterial';
+
+interface ShowItemOperation {
+  type: 'show';
+}
+
+interface HideItemOperation {
+  type: 'hide';
+}
+
+export interface OperationDefinition {
+  operation: ItemOperation;
+}
+
+export interface ChangeMaterialOperation {
+  type: 'change-material';
+  color: ColorMaterial;
+}
+
+export type ItemOperation =
+  | ShowItemOperation
+  | HideItemOperation
+  | ChangeMaterialOperation;
+
+export interface SceneItemOperations<T> {
+  materialOverride(color: ColorMaterial): T;
+  show(): T;
+  hide(): T;
+}
+
+/**
+ * A class to handle the building of operations for a scene.
+ */
+export class SceneOperationBuilder
+  implements SceneItemOperations<SceneOperationBuilder> {
+  private operations: OperationDefinition[] = [];
+
+  /**
+   * Constructs the scene operations and returns a definition describing each
+   * operation.
+   */
+  public build(): OperationDefinition[] {
+    return this.operations.concat();
+  }
+
+  public materialOverride(color: ColorMaterial): SceneOperationBuilder {
+    return this.operation({ type: 'change-material', color });
+  }
```

Comment:
Since the types in `queries.ts` are immutable, I think it'd be good to keep things consistent make this builder immutable as well, e.g. return a new `SceneOperationBuilder` with the concated operation.

Reply:
Not sure what you're implying here. If I do a new `SceneOperationBuilder().materialOverride()` this will be a recursive call.